src/dune_pkg/opam_solver.ml

```diff
@@ -177,19 +183,82 @@ module Context_for_dune = struct
+    in
+    don't_avoid @ avoid
+  ;;
+
+  (* Move all packages with a version number greater than the latest
+     released version of ocaml-base-compiler to the end of the list,
+     otherwise preserving the order of the list. If the latest
+     released version of ocaml-base-compiler is not known, return the
+     list unchanged. *)
+  let move_packages_newer_than_latest_released_base_compiler_to_end t packages =
+    match t.latest_released_base_compiler_version with
+    | Some latest_released_base_compiler_version ->
+      let latest_released_base_compiler_version =
+        Package_version.to_opam_package_version latest_released_base_compiler_version
+      in
+      let earlier_or_equal, later =
+        List.partition packages ~f:(fun package ->
+          let package_version =
+            OpamFile.OPAM.version (Resolved_package.opam_file package)
+          in
+          match
+            opam_version_compare t package_version latest_released_base_compiler_version
+          with
+          | Lt -> false
+          | Eq | Gt -> true)
+      in
+      earlier_or_equal @ later
+    | None -> packages
+  ;;
+
   let repo_candidate t name =
     let+ resolved = Opam_repo.load_all_versions t.repos name in
     let available =
-      OpamPackage.Version.Map.values resolved
-      (* This sort is not strictly necessary. The values returned from the map
-         are already sorted in ascending order. So it would be enough to just
-         reverse this list if we want the highest versions first. We leave the
-         sorting for clarity. *)
-      |> List.sort ~compare:(fun p1 p2 ->
-        opam_version_compare
-          t
-          (Resolved_package.opam_file p1)
-          (Resolved_package.opam_file p2))
+      let packages_sorted_by_version =
+        OpamPackage.Version.Map.values resolved
+        (* This sort is not strictly necessary. The values returned from the map
+           are already sorted in ascending order. So it would be enough to just
+           reverse this list if we want the highest versions first. We leave the
+           sorting for clarity. *)
+        |> List.sort ~compare:(fun p1 p2 ->
+          opam_version_compare
+            t
+            (Resolved_package.opam_file p1 |> OpamFile.OPAM.version)
+            (Resolved_package.opam_file p2 |> OpamFile.OPAM.version))
+      in
+      (if Package_name.(equal (of_opam_package_name name) (of_string "ocaml"))
+       then
+         (* This is a hack to avoid versions of the the ocaml package
+            which are later than the latest release of
+            ocaml-base-compiler. It's common for the latest version of
+            the ocaml opam package to be greater than the latest
+            released ocaml-base-compiler, in which case the ocaml
+            package will depend development releases of the underlying
+            compiler, such as +trunk releases of
+            ocaml-variants. Development releases of packages are
```

Comment:
Isn't this PR implementing support for the avoid-version flag? That is, it's taking this flag into account when returning the list of available versions.

Reply:
No not completely. After this change the solver will respect `avoid-version` between different versions a given package (e.g. avoiding beta releases of the compiler) however it will not respect `avoid-version` between different packages in general. If `foo` depends on the disjunction `bar | baz` and `bar` is `avoid-version` while `baz` is not, `bar` may still be chosen to satisfy the constraint.